Clearly separate sharing and non-sharing implementation of irept

romainbrenguier opened the pull request:
The current implementation of irept contains preprocessor instructions which makes it difficult to read, in particular with regards to SHARING.
This pull request clearly separates the sharing and non-sharing implementation, and inheritance is used for easily switching between the two and not duplicate code.
This results in more structured code, and I think much easier to read, understand.
Ultimately this could allow having sharing and none sharing structures in the same compiled binary (for instance we could experiment with using sharing for exprt but not typet, which is impossible for now).  

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Untangle dt and irept

We extract the dt class from irept, rename it to tree_nodet and make
the tree type a template argument.
This exposes more clearly how they depend on each other, and makes it
easier to reason about them independently.
In particular we now see that the node doesn't need to know much about
the tree except that it can be stored in the vector or a map.

romainbrenguier committed:
Template argument to decide on sharing

This replaces preprocessor instructions, making the code a bit clearer
and the node class more reusable.
In particular we can use both the sharing and non sharing implementation
in the same compiled binary. This will be useful when we separate the
sharing and non sharing implementations of irept.

romainbrenguier committed:
Extract a sharing_treet class from irept

The use of macros in the irept class make its difficult to understand
and prevent to use both possibilities in the same compilation.
This commit extracts from the irept class the part of the code that only
exists when SHARING is defined, which results in a simple, self
contained class.

romainbrenguier committed:
Extract a non_sharing_treet class from irept

This avoids having #ifdef SHARING and similar preprocessor instruction
in the middle of the code, and only keep one such #ifdef when irept
chooses to inherit from sharing_treet or non_sharing_treet.
This was obtained by extracting from irept, the part of the code which
would only apply when SHARING is not defined.
This results in a simple self contained class, similar to the
sharing_treet class.

tautschnig commented on src/util/irep.h:
```diff
@@ -104,6 +96,87 @@ const irept &get_nil_irep();
+  // memory and increase efficiency.
+
+#ifdef NAMED_SUB_IS_FORWARD_LIST
+  typedef std::forward_list<std::pair<irep_namet, irept>> named_subt;
```
Please make sure you also test this with `NAMED_SUB_IS_FORWARD_LIST` being set. (Which obviously hasn't been done.)

romainbrenguier replied:
Yes, I guess that's a good occasion to add tests for that.

tautschnig reviewed (commented):
I'd be curious what the profile for code using these changes vs. current develop looks like. Having `detach` not inlined was very helpful in understanding where we are failing to do perfect forwarding. Also, performance benchmarking is an absolute must for this sort of change.

romainbrenguier commented:
@tautschnig 
> Having `detach` not inlined was very helpful in understanding where we are failing to do perfect forwarding

I'm not sure what you mean, if you think it is important then could you detail that. Also I don't think the compiler would inline `detach` here.

owen-mc-diffblue commented on src/util/irep.h:
```diff
@@ -104,6 +96,87 @@ const irept &get_nil_irep();
+  unsigned ref_count = 1;
+#endif
+
+  /// This irep_idt is the only place to store data in an irep, other than
```
"the only place to store data in an irep" -> "the only place to store data in a tree node"

owen-mc-diffblue replied:
Remove ", other than the mere nesting structure", as it isn't relevant to `tree_nodet`. You could have a similar comment in `irept` somewhere if you want. (I added this comment a while ago because I found `irept` so hard to understand, so I approve of you making it clearer.)

owen-mc-diffblue commented on src/util/irep.h:
```diff
@@ -344,57 +387,29 @@ typename sharing_treet<derivedt>::dt sharing_treet<derivedt>::empty_d;
-  // use std::forward_list or std::vector< unique_ptr<T> > to save
-  // memory and increase efficiency.
-
+  using supert = tree_implementationt;
```
I think it would be slightly more consistent to call it `baset` rather than `supert`. `baset` is used in various other places in the codebase.

owen-mc-diffblue reviewed (approved):
I think this is cleaner. Obviously we should run performance benchmarks to check it doesn't have a negative impact.

romainbrenguier committed:
Add travis build with NAMED_SUB_FORWARD_LIST

We add this step to check that CBMC compiles fine with this flag, and
run relevant unit tests for irept.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b74f88b85036ceb07c80b53f017283b7b333d15a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100806145

romainbrenguier commented:
I ran this branch on my usual 70 methods:
![perf_out](https://user-images.githubusercontent.com/11854194/52770983-8df74e80-302c-11e9-8ef9-c26a1baed5ea.png)
on average it was 0.6% slower, and stopping after symex (--show-vcc --verbosity 0):
![perf_symex](https://user-images.githubusercontent.com/11854194/52771028-ad8e7700-302c-11e9-824d-fbce9c5281f9.png)
on average it was 3% faster.

I don't think the difference are significative, and I haven't seen any big drop in performances if that was a worry.

smowton commented on src/util/irep.h:
```diff
@@ -121,9 +121,9 @@ class tree_nodet : public ref_count_ift<sharing>
-// named_subt has to provide stable references; with C++11 we could
-// use std::forward_list or std::vector< unique_ptr<T> > to save
-// memory and increase efficiency.
+  // named_subt has to provide stable references; with C++11 we could
```
While you're here: this comment is now out of date (we require C++11 unconditionally)

smowton commented on src/util/irep.h:
```diff
@@ -242,26 +248,175 @@ class irept
+    write().data = _id;
+  }
+
+  non_sharing_treet(
```
Why does `non_sharing_treet` have different constructors? Surely the same take-by-value-and-move concept applies here as does for `sharing_treet`?

smowton reviewed (approved):
I like it, couple of comments

romainbrenguier committed:
Improve non_sharing_treet constructors

Taking the parameters by copy and moving them, saves some copies, in
particular we can move arguments to the constructor.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ff9d6e5ebac5848bdfb97c5bb21e2e65842ab04c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100912923
